# <a id='toc1_'></a>[**Data Wrangling and Cleaning Template**](#toc0_)

Name  
Topic  
email  
June 4th, 2023  


**Table of contents**<a id='toc0_'></a>    
- [**Cleaning Template**](#toc1_)    
  - [**Load**](#toc1_1_)    
  - [**Assess**](#toc1_2_)    
    - [Visual Assessment](#toc1_2_1_)    
    - [Programmatic Assessment](#toc1_2_2_)    
    - [List of Issues to Resolve](#toc1_2_3_)    
      - [High Level Tips Cheatsheet below for reference](#toc1_2_3_1_)    
  - [**Clean**](#toc1_3_)    
    - [Issue #1:](#toc1_3_1_)    
      - [Define:](#toc1_3_1_1_)    
      - [Code](#toc1_3_1_2_)    
      - [Test](#toc1_3_1_3_)    
    - [Issue #2:....](#toc1_3_2_)    
      - [Define:](#toc1_3_2_1_)    
      - [Code](#toc1_3_2_2_)    
      - [Test](#toc1_3_2_3_)    

<!-- vscode-jupyter-toc-config
	numbering=false
	anchor=true
	flat=false
	minLevel=1
	maxLevel=6
	/vscode-jupyter-toc-config -->
<!-- THIS CELL WILL BE REPLACED ON TOC UPDATE. DO NOT WRITE YOUR TEXT IN THIS CELL -->

## <a id='toc1_1_'></a>[**Load**](#toc0_)


These are the necessary libraries to import.

In [14]:
# Import all libraries and print versions
import pandas as pd
import numpy as np
import matplotlib as mpl
import matplotlib.pyplot as plt
import seaborn as sns

print(f"Pandas version: {pd.__version__}")
print(f"Numpy version: {np.__version__}")
print(f"Seaborn version: {sns.__version__}")
print(f"Matplotlib version: {mpl.__version__}")

# Set the random.seed(42)
np.random.seed(42)


Pandas version: 2.0.2
Numpy version: 1.23.4
Seaborn version: 0.12.2
Matplotlib version: 3.6.2


This is a demo dataset for this template. To use just make sure to replace the df with your own dataset.


In [15]:
# Load the built-in Titanic dataset from Seaborn library
df = sns.load_dataset('titanic')

# Display the first 5 rows of the dataset
df.head()


survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

## <a id='toc1_2_'></a>[**Assess**](#toc0_)

### <a id='toc1_2_1_'></a>[Visual Assessment](#toc0_)

In [16]:
# Display the first 5 rows of the DataFrame
display(df.head())

# Display the last 5 rows of the DataFrame
display(df.tail())

# Display a random sample of 5 rows from the DataFrame
display(df.sample(5))


survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

survived  pclass     sex   age  sibsp  parch   fare embarked   class  \
886         0       2    male  27.0      0      0  13.00        S  Second   
887         1       1  female  19.0      0      0  30.00        S   First   
888         0       3  female   NaN      1      2  23.45        S   Third   
889         1       1    male  26.0      0      0  30.00        C   First   
890         0       3    male  32.0      0      0   7.75        Q   Third   

       who  adult_male deck  embark_town alive  alone  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
709         1       3    male   NaN      1      1  15.2458        C   Third   
439         0       2    male  31.0      0      0  10.5000        S  Second   
840         0       3    male  20.0      0      0   7.9250        S   Third   
720         1       2  female   6.0      0      1  33.0000        S  Second   
39          1       3  female  14.0      1      0  11.2417        C   Third   

       who  adult_male deck  embark_town alive  alone  
709    man        True  NaN    Cherbourg   yes  False  
439    man        True  NaN  Southampton    no   True  
840    man        True  NaN  Southampton    no   True  
720  child       False  NaN  Southampton   yes  False  
39   child       False  NaN    Cherbourg   yes  False

### <a id='toc1_2_2_'></a>[Programmatic Assessment](#toc0_)

In [17]:
# Display a summary of the DataFrame including the index dtype and column dtypes, 
# non-null values and memory usage.
df.info()

# Generate descriptive statistics that summarize the central tendency, dispersion and shape of 
# a dataset's distribution, excluding NaN values. Transpose the output for better readability.
df.describe().T


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.7+ KB


count       mean        std   min      25%      50%   75%       max
survived  891.0   0.383838   0.486592  0.00   0.0000   0.0000   1.0    1.0000
pclass    891.0   2.308642   0.836071  1.00   2.0000   3.0000   3.0    3.0000
age       714.0  29.699118  14.526497  0.42  20.1250  28.0000  38.0   80.0000
sibsp     891.0   0.523008   1.102743  0.00   0.0000   0.0000   1.0    8.0000
parch     891.0   0.381594   0.806057  0.00   0.0000   0.0000   0.0    6.0000
fare      891.0  32.204208  49.693429  0.00   7.9104  14.4542  31.0  512.3292

**Null Value Assessment**

In [18]:
# Print the number of missing values in each column
print("Number of Null Values in Each Column:\n", df.isna().sum())

# Display the first 5 rows where 'deck' column is missing
df[df.deck.isna()].head()



Number of Null Values in Each Column:
 survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64


survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
2         1       3  female  26.0      0      0   7.9250        S  Third   
4         0       3    male  35.0      0      0   8.0500        S  Third   
5         0       3    male   NaN      0      0   8.4583        Q  Third   
7         0       3    male   2.0      3      1  21.0750        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
2  woman       False  NaN  Southampton   yes   True  
4    man        True  NaN  Southampton    no   True  
5    man        True  NaN   Queenstown    no   True  
7  child       False  NaN  Southampton    no  False

**Duplicate Value Assessment**

In [19]:
# Print the number of duplicated rows
print("Number of Duplicate Rows:", df.duplicated().sum())

# Display the first 5 rows that are duplicated
df[df.duplicated()].head()


Number of Duplicate Rows: 107


survived  pclass     sex  age  sibsp  parch    fare embarked  class  \
47         1       3  female  NaN      0      0  7.7500        Q  Third   
76         0       3    male  NaN      0      0  7.8958        S  Third   
77         0       3    male  NaN      0      0  8.0500        S  Third   
87         0       3    male  NaN      0      0  8.0500        S  Third   
95         0       3    male  NaN      0      0  8.0500        S  Third   

      who  adult_male deck  embark_town alive  alone  
47  woman       False  NaN   Queenstown   yes   True  
76    man        True  NaN  Southampton    no   True  
77    man        True  NaN  Southampton    no   True  
87    man        True  NaN  Southampton    no   True  
95    man        True  NaN  Southampton    no   True

### <a id='toc1_2_3_'></a>[List of Issues to Resolve](#toc0_)

List of Quality or Tidiness Issues found to Fix. Some issues may be found later during the EDA process. Add them here and clean while mentioning they were found later.


**Example of Quality Issues**:
1. Deck, sex, and class have several missing values. - Completeness  
   - Check to see if there is some commonality with the missing values. Otherwise if it is not a large part of the dataset (>3%), consider deleting incomplete data. 

**Example of Tidiness Issues**:   

1. There are 107 duplicate rows. - Rows   
   - Check to see if there is some commonality with the duplicate values. Otherwise if it is not a large part of the dataset (>3%), consider deleting duplicate data. 


#### <a id='toc1_2_3_1_'></a>[High Level Tips Cheatsheet below for reference](#toc0_)

**Principles of High Quality Data**:
1. **Completeness**: Do we have all of the records that we should?
2. **Validity**: We have the records, but they're not valid, i.e., they don't conform to a defined schema, also known as a defined set of rules for data.
3. **Accuracy**: Inaccurate data is wrong data that is valid. It adheres to the defined schema, but it is still incorrect.
4. **Consistency**: Inconsistent data is both valid and accurate, but there are multiple correct ways of referring to the same thing.


**Principles of Tidy Data**:   
- Tidy data is organized with three qualities in mind:
    - **Columns:** Each variable forms a column.
    - **Rows:** Each observation forms a row.
    - **Tables:** Each type of observational unit forms a table.


**Ideal Order for Addressing Issues**:

 1. **Completeness issues** or **Fix Missing Data**: It's important to do this upfront so that subsequent data cleaning will not have to be repeated.
 2. **Tidiness Issues**: Tidy datasets with data quality issues are almost always easier to clean than untidy datasets with the same issues.
 3. **Quality Control**: Address the remaining validity, accuracy, and consistency issues in that order.

## <a id='toc1_3_'></a>[**Clean**](#toc0_)

This is where you will clean the issues you listed above. You will also want to make sure you document any changes you make to the data here. You can do this by creating a copy of the original dataframes and then cleaning the copy, or you can create a new dataframe for each cleaned dataframe. 

### <a id='toc1_3_1_'></a>[Issue #1:](#toc0_)

#### <a id='toc1_3_1_1_'></a>[Define:](#toc0_)


The first issues I wanted to address were all the Completeness issues. 

**Quality**:
1. Deck, sex, and class have several missing values. - Completeness  
   - Check to see if there is some commonality with the missing values. Otherwise if it is not a large part of the dataset (>3%), consider deleting incomplete data. 


#### <a id='toc1_3_1_2_'></a>[Code](#toc0_)

In [20]:
# Cleaning code here:
null_proportion = df.isna().mean().max()
print(f'Null Value Maximum proportion: {null_proportion * 100:.2f}%')

# If the proportion of duplicates is small, consider removing them
if null_proportion < 0.03:
    df = df.dropna()


Null Value Maximum proportion: 77.22%


#### <a id='toc1_3_1_3_'></a>[Test](#toc0_)

In [21]:
# Print the number of missing values in each column
print("Number of Null Values in Each Column:\n", df.isna().sum())

# Display the first 5 rows where 'deck' column is missing
df[df.deck.isna()].head()

Number of Null Values in Each Column:
 survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64


survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
2         1       3  female  26.0      0      0   7.9250        S  Third   
4         0       3    male  35.0      0      0   8.0500        S  Third   
5         0       3    male   NaN      0      0   8.4583        Q  Third   
7         0       3    male   2.0      3      1  21.0750        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
2  woman       False  NaN  Southampton   yes   True  
4    man        True  NaN  Southampton    no   True  
5    man        True  NaN   Queenstown    no   True  
7  child       False  NaN  Southampton    no  False

### <a id='toc1_3_2_'></a>[Issue #2:....](#toc0_)

#### <a id='toc1_3_2_1_'></a>[Define:](#toc0_)


Second issue I wanted to address was the Tidiness issues of Duplicate Rows.

1. There are 107 duplicate rows. - Rows   
   - Check to see if there is some commonality with the duplicate values. Otherwise if it is not a large part of the dataset (>3%), consider deleting duplicate data. 


#### <a id='toc1_3_2_2_'></a>[Code](#toc0_)

In [22]:
# Display the first 5 rows that are duplicated
df[df.duplicated()].head()


survived  pclass     sex  age  sibsp  parch    fare embarked  class  \
47         1       3  female  NaN      0      0  7.7500        Q  Third   
76         0       3    male  NaN      0      0  7.8958        S  Third   
77         0       3    male  NaN      0      0  8.0500        S  Third   
87         0       3    male  NaN      0      0  8.0500        S  Third   
95         0       3    male  NaN      0      0  8.0500        S  Third   

      who  adult_male deck  embark_town alive  alone  
47  woman       False  NaN   Queenstown   yes   True  
76    man        True  NaN  Southampton    no   True  
77    man        True  NaN  Southampton    no   True  
87    man        True  NaN  Southampton    no   True  
95    man        True  NaN  Southampton    no   True

In [23]:
# Check the proportion of duplicate rows
duplicate_proportion = df.duplicated().mean()
print(f'Duplicate proportion: {duplicate_proportion * 100:.2f}%')


Duplicate proportion: 12.01%


In [24]:
# If the proportion of duplicates is small, consider removing them
if duplicate_proportion < 0.03:
    df = df.drop_duplicates()


#### <a id='toc1_3_2_3_'></a>[Test](#toc0_)

In [25]:
# Print the number of duplicated rows
print("Number of Duplicate Rows:", df.duplicated().sum())

# Display the first 5 rows that are duplicated
df[df.duplicated()].head()


Number of Duplicate Rows: 107


survived  pclass     sex  age  sibsp  parch    fare embarked  class  \
47         1       3  female  NaN      0      0  7.7500        Q  Third   
76         0       3    male  NaN      0      0  7.8958        S  Third   
77         0       3    male  NaN      0      0  8.0500        S  Third   
87         0       3    male  NaN      0      0  8.0500        S  Third   
95         0       3    male  NaN      0      0  8.0500        S  Third   

      who  adult_male deck  embark_town alive  alone  
47  woman       False  NaN   Queenstown   yes   True  
76    man        True  NaN  Southampton    no   True  
77    man        True  NaN  Southampton    no   True  
87    man        True  NaN  Southampton    no   True  
95    man        True  NaN  Southampton    no   True